python/cugraph/cugraph/tests/sampling/test_uniform_neighbor_sample.py

```diff
@@ -528,6 +528,206 @@ def test_uniform_neighbor_sample_empty_start_list():
+    )
+    for v in sampling_results[
+        sampling_results.hop_id == 1
+    ].sources.values_host.tolist():
+        assert v not in next_sources
+
+
+@pytest.mark.sg
+def test_uniform_neighbor_sample_unique_sources_email_eu_core():
+    el = email_Eu_core.get_edgelist()
+
+    G = cugraph.Graph(directed=True)
+    G.from_cudf_edgelist(el, source="src", destination="dst")
+
+    seeds = el.src.sample(frac=0.001).unique().sample(frac=1)
+
+    sampling_results = cugraph.uniform_neighbor_sample(
+        G,
+        seeds,
+        [5, 4, 3, 2, 1],
+        with_replacement=False,
+        with_edge_properties=True,
+        with_batch_ids=False,
+        unique_sources=True,
+    )
+
+    for hop in range(5):
+        current_sources = set(
+            sampling_results[
+                sampling_results.hop_id == hop
+            ].sources.values_host.tolist()
+        )
+        future_sources = set(
+            sampling_results[sampling_results.hop_id > hop].sources.values_host.tolist()
+        )
+
+        for s in current_sources:
+            assert s not in future_sources
+
+
+@pytest.mark.sg
+def test_uniform_neighbor_sample_carry_over_sources_basic():
+    df = cudf.DataFrame(
+        {
+            "src": [0, 4, 1, 2, 3, 5, 4, 1, 0, 6],
+            "dst": [1, 1, 2, 4, 6, 1, 5, 0, 2, 2],
+            "eid": [9, 8, 7, 6, 5, 4, 3, 2, 1, 0],
+        }
+    )
+
+    G = cugraph.MultiGraph(directed=True)
+    G.from_cudf_edgelist(df, source="src", destination="dst", edge_id="eid")
+
+    sampling_results = cugraph.uniform_neighbor_sample(
+        G,
+        cudf.DataFrame(
+            {
+                "seed": cudf.Series([0, 4, 3], dtype="int64"),
+                "batch": cudf.Series([1, 1, 1], dtype="int32"),
+            }
+        ),
+        [2, 3, 3],
+        with_replacement=False,
+        with_edge_properties=True,
+        with_batch_ids=True,
+        random_state=62,
+        carry_over_sources=True,
+    ).sort_values(by="hop_id")[["sources", "destinations", "hop_id"]]
+
+    assert (
+        len(
+            sampling_results[
+                (sampling_results.hop_id == 2) & (sampling_results.sources == 6)
+            ]
+        )
+        == 2
+    )
+
+    for hop in range(2):
+        sources_current_hop = set(
+            sampling_results[
+                sampling_results.hop_id == hop
+            ].sources.values_host.tolist()
+        )
+        sources_next_hop = set(
+            sampling_results[
+                sampling_results.hop_id == (hop + 1)
+            ].sources.values_host.tolist()
+        )
+
+        for s in sources_current_hop:
+            assert s in sources_next_hop
+
+
+@pytest.mark.sg
+def test_uniform_neighbor_sample_carry_over_sources_email_eu_core():
+    el = email_Eu_core.get_edgelist()
+
+    G = cugraph.Graph(directed=True)
+    G.from_cudf_edgelist(el, source="src", destination="dst")
+
+    seeds = el.src.sample(frac=0.001).unique().sample(frac=1)
```

Comment:
I believe you can also use [select_random_vertices](https://github.com/rapidsai/cugraph/blob/branch-23.08/python/cugraph/cugraph/structure/graph_implementation/simpleGraph.py#L674)

Reply:
fixed